File: 0525-contiguous-array/0525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        diff_index_dct = {}
        diff_index_dct[0]= -1
        max_len = runningSum = 0
        for right in range(len(nums)):
            if nums[right] == 0:
                runningSum -= 1
            else:
                runningSum += 1
            if runningSum in diff_index_dct:
                max_len = max(max_len, right - diff_index_dct[runningSum])
            else:
                diff_index_dct[runningSum] = right
        return max_len

        

File: 0525-contiguous-array/0525-contiguous-array.rs
use std::collections::HashMap;
impl Solution {
    pub fn find_max_length(nums: Vec<i32>) -> i32 {
        let mut diffIndexMap : HashMap<i32, i32> = HashMap::new(); // difference, index of its first occurence
        let mut runningSum = 0;
        let mut maxLength = 0;
        diffIndexMap.insert(0, -1);
        for (right, &num) in nums.iter().enumerate(){
            if num == 0{
                runningSum -= 1;
            }else{
                runningSum += 1;
            }
            if let Some(&diff) = diffIndexMap.get(&runningSum){
                maxLength = maxLength.max(right as i32 - diff);
            }else{
                diffIndexMap.insert(runningSum, right as i32);
            }
        }
        maxLength
    }
}